internal/backend/runtime/omni/controllers/reporting/reporting.go

```diff
@@ -0,0 +1,204 @@
+
+package reporting
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/controller/generic"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/stripe/stripe-go/v74"
+	"github.com/stripe/stripe-go/v74/subscriptionitem"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+)
+
+var (
+	// DebounceDuration defines how long to wait after the last event before processing updates.
+	DebounceDuration = 10 * time.Minute
+)
+
+// StripeMetricsReporterControllerName is the name of the StripeMetricsReporterController.
+const StripeMetricsReporterControllerName = "StripeMetricsReporterController"
+
+// StripeMetricsReporterController ...
+type StripeMetricsReporterController struct {
+	generic.NamedController
+}
+
+// NewStripeMetricsReporterController initializes StripeMetricsReporterController.
+func NewStripeMetricsReporterController() *StripeMetricsReporterController {
+	return &StripeMetricsReporterController{
+		NamedController: generic.NamedController{
+			ControllerName: StripeMetricsReporterControllerName,
+		},
+	}
+}
+
+// Settings implements controller.QController interface.
+func (ctrl *StripeMetricsReporterController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: resources.DefaultNamespace,
+			Type:      omni.MachineStatusMetricsType,
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *StripeMetricsReporterController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+func (ctrl *StripeMetricsReporterController) Run(ctx context.Context, r controller.Runtime, _ *zap.Logger) error {
+	// Set your Stripe API key from the environment variable.
+	stripe.Key = os.Getenv("STRIPE_API_KEY")
+	if stripe.Key == "" {
+		return fmt.Errorf("STRIPE_API_KEY environment variable not set")
+	}
+
+	// Retrieve Subscription Item ID from environment variables or configuration.
+	subscriptionItemID := os.Getenv("STRIPE_SUBSCRIPTION_ITEM_ID")
+	if subscriptionItemID == "" {
+		return fmt.Errorf("STRIPE_SUBSCRIPTION_ITEM_ID environment variable not set")
+	}
+
+	// Channel to send count changes to the debouncer.
+	countChan := make(chan uint32, 100) // Buffered to handle bursts.
+
+	// Start the debouncer goroutine.
+	go debouncer(ctx, countChan, subscriptionItemID)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+			metrics, err := safe.ReaderGetByID[*omni.MachineStatusMetrics](ctx, r, omni.MachineStatusMetricsID)
+			if err != nil {
+				if state.IsNotFoundError(err) {
+					continue
+				}
+
+				return err
+			}
+
+			count := metrics.TypedSpec().Value.RegisteredMachinesCount
+			countChan <- count
```

Comment:
Maybe protect this with `select` from blocking by selecting on `ctx.Done()` and this channel send?